src/main/java/io/cdap/plugin/http/sink/batch/MessageBuffer.java

```diff
@@ -0,0 +1,203 @@
+  private static final Map<String, String> contentTypeMap = new HashMap<String, String>() {{
+    put("json", "application/json");
+    put("form", "application/x-www-form-urlencoded");
+    put("custom", "text/plain");
+  }};
+
+  public MessageBuffer(
+          String messageFormat, String jsonBatchKey, Boolean shouldWriteJsonAsArray,
+          String delimiterForMessages, String charset, String customMessageBody
+  ) {
+    this.jsonBatchKey = jsonBatchKey;
+    this.delimiterForMessages = delimiterForMessages;
+    this.charset = charset;
+    this.shouldWriteJsonAsArray = shouldWriteJsonAsArray;
+    this.customMessageBody = customMessageBody;
+    this.buffer = new ArrayList<>();
+    switch (messageFormat.toLowerCase()) {
+      case "json":
+        messageFormatter = this::formatAsJsonNoException;
+        contentType = contentTypeMap.get("json");
+        break;
+      case "form":
+        messageFormatter = this::formatAsForm;
+        contentType = contentTypeMap.get("form");
+        break;
+      case "custom":
+        messageFormatter = this::formatAsCustom;
+        contentType = contentTypeMap.get("custom");
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid message format: " + messageFormat);
+    }
+  }
+
+  public void add(StructuredRecord record) {
+    buffer.add(record);
+  }
+
+  public void clear() {
+    buffer.clear();
+  }
+
+  public int size() {
+    return buffer.size();
+  }
+
+  public boolean isEmpty() {
+    return buffer.isEmpty();
+  }
+
+  public String getContentType() {
+    return contentType;
+  }
+
+  public String getMessage() throws IOException {
+    return messageFormatter.apply(buffer);
+  }
+
+  private String formatAsJsonNoException(List<StructuredRecord> buffer) {
+    try {
+      return formatAsJson(buffer);
+    } catch (IOException e) {
+      throw new IllegalStateException("Error formatting JSON message. Reason: " + e.getMessage(), e);
+    }
+  }
+
+  private String formatAsJson(List<StructuredRecord> buffer) throws IOException {
+    boolean useJsonBatchKey = !Strings.isNullOrEmpty(jsonBatchKey);
+    if (!shouldWriteJsonAsArray) {
+      return String.join(delimiterForMessages, getBufferAsJsonList());
+    }
+    if (useJsonBatchKey) {
+      // A new StructuredRecord is created with the jsonBatchKey as the
+      // field name and the array of records as the value
+      Schema bufferRecordSchema = buffer.get(0).getSchema();
+      Schema bufferRecordArraySchema = Schema.arrayOf(bufferRecordSchema);
+      Schema wrappedMessegeSchema = Schema.recordOf("wrapper",
+              Schema.Field.of(jsonBatchKey, bufferRecordArraySchema));
+      StructuredRecord wrappedMessageRecord = StructuredRecord.builder(wrappedMessegeSchema)
+              .set(jsonBatchKey, buffer).build();
+      return StructuredRecordStringConverter.toJsonString(wrappedMessageRecord);
+    } else {
+      return "[" + String.join(",", getBufferAsJsonList()) + "]";
```

Comment:
still have the comment about doing this in one loop instead of two. For example, can change getBufferAsJsonList() to take a StringBuilder as an argument and append to it as needed.

Reply:
This will do the work in one loop
```java
private String getBufferAsJsonList() throws IOException {
    StringBuilder sb = new StringBuilder();
    int size = buffer.size();
    String delimiter = shouldWriteJsonAsArray ? "," : delimiterForMessages;
    if (shouldWriteJsonAsArray) {
      sb.append("[");
    }
    for (StructuredRecord record : buffer) {
      sb.append(StructuredRecordStringConverter.toJsonString(record));
      if (--size > 0) {
        sb.append(delimiter);
      }
    }
    if (shouldWriteJsonAsArray) {
      sb.append("]");
    }
    return sb.toString();
  }
  ```